src/lib/CollisionPrevention/CollisionPrevention.cpp

```diff
@@ -69,104 +65,83 @@ bool CollisionPrevention::initializeSubscriptions(SubscriptionArray &subscriptio
 
 		for (int i = 0; i < distances_array_size; i++) {
+
 			//determine if distance bin is valid and contains a valid distance measurement
 			if (obstacle_distance.distances[i] < obstacle_distance.max_distance &&
 			    obstacle_distance.distances[i] > obstacle_distance.min_distance && i * obstacle_distance.increment < 360) {
 				float distance = obstacle_distance.distances[i] / 100.0f; //convert to meters
 				float angle = math::radians((float)i * obstacle_distance.increment);
 
-				//calculate normalized velocity reductions
-				float vel_lim_x = (max_detection_distance - distance) / (max_detection_distance - _param_mpc_col_prev_d.get()) * cos(
-							  angle);
-				float vel_lim_y = (max_detection_distance - distance) / (max_detection_distance - _param_mpc_col_prev_d.get()) * sin(
-							  angle);
+				//max velocity squared in current bin direction
+				float vel_max_sqrd = math::max(0.f, max_acc * (distance - _param_mpc_col_prev_d.get()));
+
+				//split current setpoint into parallel and orthogonal components with respect to the current bin direction
+				Vector2f bin_direction = {cos(angle), sin(angle)};
+				Vector2f orth_direction = {-sin(angle), cos(angle)};
+				float sp_parallel = setpoint_sqrd.dot(bin_direction);
+				float sp_orth = setpoint_sqrd.dot(orth_direction);
+
+				//limit the setpoint to respect vel_max by subtracting from the parallel component
+				if (sp_parallel > vel_max_sqrd) {
+					Vector2f setpoint_temp = setpoint_sqrd - (sp_parallel - vel_max_sqrd) * bin_direction;
 
-				if (vel_lim_x > 0 && vel_lim_x > _move_constraints_x_normalized(1)) { _move_constraints_x_normalized(1) = vel_lim_x; }
+					//limit sliding angle
+					float angle_diff_temp_orig = acos(setpoint_temp.dot(setpoint) / (setpoint_temp.norm() * setpoint.norm()));
+					float angle_diff_temp_bin = acos(setpoint_temp.dot(bin_direction) / (setpoint_temp.norm() * bin_direction.norm()));
 
-				if (vel_lim_y > 0 && vel_lim_y > _move_constraints_y_normalized(1)) { _move_constraints_y_normalized(1) = vel_lim_y; }
+					if (angle_diff_temp_orig > max_slide_angle_rad && setpoint_temp.norm() > 0.001f) {
+						float angle_temp_bin_cropped = angle_diff_temp_bin - (angle_diff_temp_orig - max_slide_angle_rad);
+						float orth_len = vel_max_sqrd * tan(angle_temp_bin_cropped);
 
-				if (vel_lim_x < 0 && -vel_lim_x > _move_constraints_x_normalized(0)) { _move_constraints_x_normalized(0) = -vel_lim_x; }
+						if (sp_orth > 0) {
+							setpoint_temp = vel_max_sqrd * bin_direction + orth_len * orth_direction;
 
-				if (vel_lim_y < 0 && -vel_lim_y > _move_constraints_y_normalized(0)) { _move_constraints_y_normalized(0) = -vel_lim_y; }
+						} else {
+							setpoint_temp = vel_max_sqrd * bin_direction - orth_len * orth_direction;
+						}
+					}
+
+					setpoint_sqrd = setpoint_temp;
+				}
 			}
 		}
 
+		//take the squared root
+		if (setpoint_sqrd.norm() > 0.001f) {
+			setpoint = setpoint_sqrd.normalized() * std::sqrt(setpoint_sqrd.norm());
```

Comment:
This formula expands to `x / norm(x) * sqrt(norm(x))`, so those `norm(x)`s simplify out. So this just needs to be

`setpoint = setpoint_sqrd / std::sqrt(setpoint_sqrd.norm());`